### PR TITLE
Support gathering additional arbitrary properties from referenced projects

### PR DESCRIPTION
**Pull request by dsplaisted**
This allows additional properties to be gathered from project references via the project reference protocol.  This is in order to support generating an error when a self-contained Executable references a non-SelfContained Executable, or vice versa, as described in dotnet/sdk#15117.

The referenced project can declare what additional properties should be gathered with `AdditionalTargetFrameworkInfoProperty` items:

```xml
  <ItemGroup>
    <AdditionalTargetFrameworkInfoProperty Include="SelfContained"/>
    <AdditionalTargetFrameworkInfoProperty Include="_IsExecutable"/>
  </ItemGroup>
```

These items will then be included in the `AdditionalPropertiesFromProject` metadata on the `_MSBuildProjectReferenceExistent` items.  This value will have `PropertyName=PropertyValue` combinations joined by semicolons, and those sets of properties for the different `TargetFramework` values will be joined by double semicolons.  For example, a project multitargeted to two TargetFrameworks may have the following for the `AdditionalPropertiesFromProject` metadata:

```
SelfContained=true;_IsExecutable=true;;SelfContained=false;_IsExecutable=true
```

Finding the right set of properties from the referenced project will required looking up the index of the `NearestTargetFramework` in the `TargetFrameworks`, and using that index to select the set of properties in the `AdditionalPropertiesFromProject` metadata.  For example:

```C#
string nearestTargetFramework = project.GetMetadata("NearestTargetFramework");
int targetFrameworkIndex = project.GetMetadata("TargetFrameworks").Split(';').ToList().IndexOf(nearestTargetFramework);
string projectAdditionalPropertiesMetadata = project.GetMetadata("AdditionalPropertiesFromProject").Split(new[] { ";;" }, StringSplitOptions.None)[targetFrameworkIndex];
Dictionary<string, string> projectAdditionalProperties = new(StringComparer.OrdinalIgnoreCase);
foreach (var propAndValue in projectAdditionalPropertiesMetadata.Split(';'))
{
    var split = propAndValue.Split('=');
    projectAdditionalProperties[split[0]] = split[1];
}
```

This is implemented in dotnet/sdk#15134.

If anyone has suggestions for a better separator than a double semicolon to separate the list of property/values for each TargetFramework, let me know.

### CODE REVIEW

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 1]
```diff
@@ -1730,6 +1730,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
         <TargetFrameworks>@(_TargetFrameworkInfo)</TargetFrameworks>
         <TargetFrameworkMonikers>@(_TargetFrameworkInfo->'%(TargetFrameworkMonikers)')</TargetFrameworkMonikers>
         <TargetPlatformMonikers>@(_TargetFrameworkInfo->'%(TargetPlatformMonikers)')</TargetPlatformMonikers>
+        <AdditionalPropertiesFromProject>@(_TargetFrameworkInfo->'%(AdditionalPropertiesFromProject)', ';;')</AdditionalPropertiesFromProject>
```
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:1733]: What does including `, ';;'` do? My test* didn't look like it did anything.

\* Looked at J's metadata after:
```
<I Include=""initial"">
<M>hi=one;not=two;;hi=three;or=four</M>
</I>
<J Include=""@(I)"">
<ExtraProp>@(I->'%(M)', ';;')</ExtraProp>
<ExtraPropTwo>@(I->'%(M)')</ExtraPropTwo>
</J>
```

**Review by Forgind — COMMENTED** [review]
I'm not a fan of relying on `;;` as a separator, since a lot of people add `;` to the beginning or end of lists if they aren't sure whether a property is empty or not like $(P);$(Q);(R) if P and Q happen to be empty, and that would confuse parsing it. Perhaps something a little more complicated like (TF1);;(TF2);;(TF3)? It's also a bit annoying to have to do so much parsing. Makes me wish we had item metadata metadata 😉
How annoying would it be to have a collection of Items like:
<ItemGroup>
<Net472AdditionalPropertiesFromProject>
<...>
</Net472AdditionalPropertiesFromProject>
<Netcoreapp3.1AdditionalPropertiesFromProject>
<...>
</Netcoreapp3.1AdditionalPropertiesFromProject>
? Is it even possible to pass that through a ProjectReference?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 1 above)
Comment (dsplaisted) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:1733]: It specifies what should be used to separate the items in the list when the list is converted to a string: https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-display-an-item-list-separated-with-commas?view=vs-2019

You would need more than one `I` item when testing it to see the difference.

**Comment by dsplaisted** [comment]
@forgind I think it will handle empty properties and empty lists of properties correctly, though I'm not 100% sure.

Something like `Net472AdditionalPropertiesFromProject` won't work, as it needs to collect properties from whatever target frameworks the referenced project defines, which is an open set.  See  dotnet/sdk#15134 for how this would be used.

**Comment by Forgind** [comment]
This was my case:
```
<PropertyGroup>
<EmptyProperty Condition="false">x</EmptyProperty>
<NotEmpty Condition="true">value</NotEmpty>
<AlsoNotEmpty Condition="true">otherVal</AlsoNotEmpty>
</PropertyGroup>
...
<PropertyGroup>
<UsesProperties>$(AlsoNotEmpty);$(EmptyProperty);$(NotEmpty)</UsesProperties>
</PropertyGroup>
...
...

and then trying to pass that to another project. I'm not sure either, mostly because I get confused sometimes by item transforms and such. As I understand it, it would accept UsesProperties as a valid property (which it is) and add it to AdditionalPropertiesFromProject, then parsing would break the property in two, get confused, and silently use it incorrectly. I could be wrong.

**Comment by dsplaisted** [comment]
@cdmihai How would this interact with static graph?

**Comment by rainersigwald** [comment]
High-level questions (this implementation looks ok but I wonder if we can improve the overall design).

1. Would this be better as one-metadatum-per-TF? With names like `AdditionalPropertiesFromProject_net472`? That'd be somewhat harder to assemble but avoids (one layer of) the delimiting problem.
2. Another option would be to pass it back as an XML blob. That's what solutions do for similar reasons.

Those are both harder to implement with MSBuild logic and would suggest writing a new `AggregateMultitargetingInformation` task, but that'd probably be easy to implement and understand.

If we stick with delimited, maybe consider using one of the fancier unicode characters like [`INFORMATION SEPARATOR TWO`](https://www.fileformat.info/info/unicode/char/001e/index.htm)?

**Comment by cdmihai** [comment]
Looks fine for static graph, as the msbuild task calling patterns haven't changed, just the data ferried by the calls, which is opaque to static graph builds.

**Comment by cdmihai** [comment]
Can you also update the [protocol doc](https://github.com/dotnet/msbuild/blob/master/documentation/ProjectReference-Protocol.md) in this PR?

**Comment by Forgind** [comment]
Team triage:
@dsplaisted
Just checking whether the feedback on this was clear—we don't think this is mergeable as-is without one of the changes rainersigwald proposed. If it's just that there are other things higher on your priority list, carry on.

**Comment by dsplaisted** [comment]
@Forgind I've updated this to use Xml to pass the data around

File: src/Tasks/CombineTargetFrameworkInfoProperties.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            XElement root = new XElement(RootElementName);
+
+            foreach (var item in PropertiesAndValues)
+            {
+                root.Add(new XElement(item.ItemSpec, item.GetMetadata("Value")));
+            }
+
+            Result = root.ToString();
+
+            return true;
```
Comment (Forgind) [inline src/Tasks/CombineTargetFrameworkInfoProperties.cs:31]: return !HasLoggedErrors? Same below.

File: src/Tasks/CombineXmlElements.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+
+namespace Microsoft.Build.Tasks
+{
+    public class CombineXmlElements : TaskExtension
+    {
+        public string RootElementName { get; set; }
```
Comment (Forgind) [inline src/Tasks/CombineXmlElements.cs:13]: Required annotations for these?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 4]
```diff
@@ -1743,12 +1751,26 @@ Copyright (C) Microsoft Corporation. All rights reserved.
   <Target Name="GetTargetFrameworksWithPlatformForSingleTargetFramework"
         Returns="@(_TargetFrameworkInfo)">
 
+    <ItemGroup>
+      <_AdditionalTargetFrameworkInfoPropertyWithValue Include="@(AdditionalTargetFrameworkInfoProperty)">
+        <Value>$(%(AdditionalTargetFrameworkInfoProperty.Identity))</Value>
```
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:1756]: I didn't realize you could have a metadata reference as an identifier for a property. Neat!

**Review by Forgind — APPROVED** [review]
Some nits, but this looks good, thanks!

**Comment by cdmihai** [comment]
Are the tests for this new flow in the sdk repo? If not, add some here?

**Comment by dsplaisted** [comment]
> Are the tests for this new flow in the sdk repo? If not, add some here?

Yes, tests for this are in dotnet/sdk#15134

**Comment by dsplaisted** [comment]
I've applied the code review feedback and retargeted this to the master branch for VS 16.10.

**Review by cdmihai — APPROVED** [review]

**Review by KirillOsenkov — APPROVED** [review]

**Comment by ericstj** [comment]
Looks like this broke WCF as it doesn't handle cases like `portable-net45+win8+wp8`

**Comment by dsplaisted** [comment]
> Looks like this broke WCF as it doesn't handle cases like `portable-net45+win8+wp8`

@ericstj Do you have more details on how it broke?  Is it something we need to fix?  I don't know if we support the old PCL TFMs anymore.

**Comment by ericstj** [comment]
I filed https://github.com/dotnet/msbuild/issues/6603 to track the regression.  It was due to an invalid XML character in the TargetFramework.  I imagine a fix would be to make sure you escape the strings which seems like something you might want to do regardless.

I was able to help WCF find a workaround since they aren't actually building for PCLs (they are just trying to use CSProj + pack to correctly represent a package which overlaps with the PCL targeting packs).  https://github.com/ericstj/wcf/commit/66e68f2da8699c913c7f3887bc9e32942f9053f1

**Pull request merged**